### PR TITLE
export aliasee if not export all and if alias is exported

### PR DESCRIPTION
**Pull request by waterlike86**
we found that when using an exports list (i.e. not export_all), the side module imports the alias. 
but just exporting the alias is not enough, we need to export the actual implementation as well (i.e. the aliasee, aliased)

### CODE REVIEW

**Comment by waterlike86** [comment]
@kripken do you think this makes sense? we were trying to obtain the list of aliases by scanning the imports and using llvm to scan the BC files, but there was always some which cannot be obtained and the BC level.
for example 
in stdlib, __ZNSt12length_errorD1Ev maps to __ZNSt11logic_errorD2Ev and this information is not available at the BC level, but only available after JSBackend. 
would you have any idea how we could get this from outside of the tool chain so we can avoid this modification?

**Comment by waterlike86** [comment]
cc @awtcode

**Comment by kripken** [comment]
I think this is right, yes. I thought we already handled it, though, but I'm not sure where.

Please add a testcase - we need one anyhow, and it will also help understand things.

**Comment by waterlike86** [comment]
@kripken added a test case

**Comment by waterlike86** [comment]
> I think this is right, yes. I thought we already handled it, though, but I'm not sure where.

actually i see in here it might have been handled for side_modules? 
https://github.com/emscripten-core/emscripten/blob/6aea2c2414f33473e725053904813ee4f1c86887/emscripten.py#L1609

**Comment by waterlike86** [comment]
> > I think this is right, yes. I thought we already handled it, though, but I'm not sure where.
> 
> @kripken actually i see in here it might have been handled for side_modules?
> [emscripten/emscripten.py](https://github.com/emscripten-core/emscripten/blob/6aea2c2414f33473e725053904813ee4f1c86887/emscripten.py#L1609)
> 
> Line 1609 in [6aea2c2](/emscripten-core/emscripten/commit/6aea2c2414f33473e725053904813ee4f1c86887)
> 
>  for k, v in metadata['aliases'].items():

File: emscripten.py [hunk 1]
```diff
@@ -874,6 +874,17 @@ def get_exported_implemented_functions(all_exported_functions, all_implemented,
     if key in all_exported_functions or export_all or (export_bindings and key.startswith('_emscripten_bind')):
       funcs.add(key)
 
+  if not export_all:
+    aliases = metadata['aliases'].keys()
+    for key in all_exported_functions:
```
Comment (kripken) [inline emscripten.py]: Is this outer loop needed? (`all_exported_functions` may be very large, which worries me)

Instead, how about
```
for name, alias in metadata['aliases']:
  if name in all_implemented:
    funcs.add(alias)
```

File: tests/test_other.py [hunk 2]
```diff
@@ -8103,6 +8103,39 @@ def test_no_legalize_js_ffi(self):
...
+      print(' '.join(cmd))
+      run_process(cmd)
+      run_process([wasm_dis, 'main.out.wasm', '-o', 'main.out.wast'])
+      text = open('main.out.wast').read()
+      # remove internal comments and extra whitespace
+      text = re.sub(r'\(;[^;]+;\)', '', text)
+      text = re.sub(r'\$var\$*.', '', text)
+      text = re.sub(r'param \$\d+', 'param ', text)
+      text = re.sub(r' +', ' ', text)
+      # print("text: %s" % text)
```
Comment (kripken) [inline tests/test_other.py]: can we remove all the text parsing here? if it is enough to just test for success in running the program, that is simpler and better.

**Comment by kripken** [comment]
I think that existing code (line 1609 you found) handles exporting from the asm.js module, but your new code here is still needed to have the functions in the final list of exports overall. To confirm, I assume the test fails without your code change?

File: emscripten.py (hunk 1 above)
Comment (waterlike86) [reply emscripten.py]: thats is actually to reduce the number of exports. we could just export all the alias like what you did, i think there arent many.

File: tests/test_other.py (hunk 2 above)
Comment (waterlike86) [reply tests/test_other.py]: sure let me do that.

**Comment by waterlike86** [comment]
> I think that existing code (line 1609 you found) handles exporting from the asm.js module, but your new code here is still needed to have the functions in the final list of exports overall. To confirm, I assume the test fails without your code change?

yes you are right. the test still fails without my code change.

File: emscripten.py (hunk 1 above)
Comment (waterlike86) [reply emscripten.py]: thats also why its only run if not export_all, so there shouldnt be that many exports i hope.

**Comment by waterlike86** [comment]
strangely i see test which seem to fail which are unrelated to my change

**Comment by kripken** [comment]
I think those errors can be ignored (should vanish with merging in latest incoming to here), except for `other.test_only_force_stdlibs` which looks like we need to figure it out here.

**Comment by waterlike86** [comment]
@kripken seems like all tests passed after i did a merge from incoming

**Comment by kripken** [comment]
Thanks @waterlike86, looks good!

**Pull request merged**